### PR TITLE
Fixes #15606: transitive dependencies missing in clearbody.

### LINKED ISSUE #15606: `clearbody` creates ill-typed terms
If I do

```coq
Goal let x := True in let y := eq_refl x : x = True in True.
Proof.
  intros. clearbody x.
```

as expected, `clearbody` tells me that I can't clear the body of `x`: `Hypothesis y depends on the body of x.`

However, in the following example

```coq
Record map (n: nat) := mk {
  rep : Type;
}.

Ltac lots_of_ltac := idtac.

Goal forall (n: nat) (mem: map n) (m: rep n mem), n = 2 -> m = m.
Proof.
  intros.
  set (Map := map n) in *.
  clearbody Map.
  lots_of_ltac.
  subst.
```

`clearbody` succeeds and creates the goal

```
  n : nat
  Map : Type
  mem : Map
  m : rep n mem
  H : n = 2
  ============================
  m = m
```

where the type of `m` does not typecheck, and later, the `subst` fails with

```
Error: Abstracting over the term "n" leads to a term
fun n0 : nat => forall m : rep n0 mem, m = m
which is ill-typed.
Reason is: Illegal application:
The term "rep" of type "forall n : nat, map n -> Type"
cannot be applied to the terms
 "n0" : "nat"
 "mem" : "Map"
The 2nd term has type "Map" which should be coercible to "map n0".
```

This is inconvenient, because
1) I thought `subst` should never fail.
2) The ill-typed term was already created by `clearbody Map`, so it's hard to debug if the source of the problem and the point where it is noticed are far apart.

And here's a more minimized example by @JasonGross:

```coq
Goal let U := Set in forall T : U, T -> True.
  intros.
  clearbody U.
```

Coq version: 8.15-rc1

### PR DESCRIPTION
**Pull request by herbelin**
**Kind:** fix

Tactic `clearbody` was missing check of transitive dependencies.

Fixes / closes #15606

I seized the opportunity to avoid computing too early the error messages possibly raised by `clearbody`, and rewrote the algorithm to stop exploring the context whenever the bodies to clear are all found.

- [x] Added / updated **test-suite**.
- [x] Added **changelog**.

### CODE REVIEW

**Comment by Alizter** [comment]
There are no tests, were you intending to add them?

**Comment by herbelin** [comment]
> There are no tests, were you intending to add them?

Hum, yes. Now done.

File: doc/changelog/04-tactics/15634-master+fix-15606-clearbody-transitive-check.rst [hunk 1]
```diff
@@ -0,0 +1,6 @@
+- **Fixed:**
+  Ill-typed goals created by :tacn:`clearbody` in the presence of
+  transitive dependencies in the body of an hypothesis
```
Comment (Alizter) [inline doc/changelog/04-tactics/15634-master+fix-15606-clearbody-transitive-check.rst:3]:
```suggestion
  Ill-typed goals created by :tacn:`clearbody` in the presence of
  transitive dependencies in the body of a hypothesis
```

**Review by Alizter — APPROVED** [review]

**Comment by Alizter** [comment]
cc @SkySkimmer Is this good for 8.15?

**Comment by ppedrot** [comment]
Bench: https://gitlab.com/coq/coq/-/jobs/2060215599

**Comment by SkySkimmer** [comment]
>cc @SkySkimmer Is this good for 8.15?

looks ok

**Comment by ppedrot** [comment]
```
┌─────────────────────────────┬─────────────────────────┬───────────────────────────────────────┬───────────────────────────────────────┬─────────────────────────┬─────────────────┐
│                             │      user time [s]      │              CPU cycles               │           CPU instructions            │  max resident mem [KB]  │   mem faults    │
│                             │                         │                                       │                                       │                         │                 │
│                package_name │     NEW      OLD  PDIFF │            NEW             OLD  PDIFF │            NEW             OLD  PDIFF │     NEW      OLD  PDIFF │ NEW  OLD  PDIFF │
├─────────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│        coq-metacoq-template │   72.46    73.82  -1.84 │   323522355630    330469986370  -2.10 │   492295627835    492249979607   0.01 │  693508   698428  -0.70 │   0    0    nan │
│                   coq-color │  206.84   210.09  -1.55 │   936097952718    952132788829  -1.68 │  1328660070741   1328526682402   0.01 │ 1142860  1143076  -0.02 │   0    0    nan │
│                coq-coqprime │   58.54    59.45  -1.53 │   266902781475    270512051391  -1.33 │   445511429314    445530179120  -0.00 │  758988   758536   0.06 │   0    0    nan │
│         coq-category-theory │ 1586.99  1611.57  -1.53 │  7265840333805   7377713128602  -1.52 │ 15271858479987  15272080776173  -0.00 │ 1473936  1474036  -0.01 │   0    0    nan │
│           coq-iris-examples │  419.58   425.75  -1.45 │  1911347613770   1938771334874  -1.41 │  2907280863354   2907496829451  -0.01 │ 1171036  1169268   0.15 │   0    0    nan │
│      coq-mathcomp-ssreflect │   25.96    26.34  -1.44 │   118155667255    119730575210  -1.32 │   149204967098    149155701897   0.03 │  531352   533940  -0.48 │   0    0    nan │
│      coq-mathcomp-odd-order │  546.50   554.46  -1.44 │  2503992322279   2540065904115  -1.42 │  4257122706674   4257025483454   0.00 │  932992   932040   0.10 │   0    0    nan │
│               coq-perennial │ 3918.45  3971.53  -1.34 │ 17888432575893  18134696748594  -1.36 │ 29798364774461  29798503270125  -0.00 │ 2601996  2597868   0.16 │   0    0    nan │
│                 coq-coqutil │   34.42    34.88  -1.32 │   154455738819    157278852949  -1.79 │   209529380202    209468864727   0.03 │  601396   601200   0.03 │   0    0    nan │
│                    coq-corn │  752.00   761.43  -1.24 │  3427863716739   3470990190139  -1.24 │  5283796890094   5283708239252   0.00 │  873508   873096   0.05 │   0    0    nan │
│                   coq-verdi │   47.88    48.44  -1.16 │   216622169142    218882740728  -1.03 │   322557080762    322599253703  -0.01 │  836956   839408  -0.29 │   0    0    nan │
│              coq-coquelicot │   33.94    34.30  -1.05 │   152924756361    154880208247  -1.26 │   202162116871    202147420221   0.01 │  747920   746360   0.21 │   4    0    nan │
│        coq-mathcomp-algebra │   63.45    64.12  -1.04 │   290237325172    292958210204  -0.93 │   398606890317    398552789649   0.01 │  556776   556340   0.08 │   0    0    nan │
│      coq-mathcomp-character │   76.52    77.25  -0.94 │   350221415863    353614871925  -0.96 │   526199848573    526202125682  -0.00 │  733080   731620   0.20 │   0    0    nan │
│       coq-mathcomp-solvable │   87.81    88.55  -0.84 │   401329107940    405200946488  -0.96 │   611654271081    611711058613  -0.01 │  658300   661472  -0.48 │   0    0    nan │
│           coq-metacoq-pcuic │  724.56   730.30  -0.79 │  3277882419743   3305454941592  -0.83 │  4739195334372   4738590537459   0.01 │ 1232856  1232552   0.02 │   0    0    nan │
│       coq-mathcomp-fingroup │   23.51    23.69  -0.76 │   107316111907    108306708327  -0.91 │   156288545754    156260019203   0.02 │  484280   485972  -0.35 │   0    0    nan │
│       coq-engine-bench-lite │  176.55   177.88  -0.75 │   765435038543    771444072625  -0.78 │  1500191764314   1500417050157  -0.02 │ 1175964  1173932   0.17 │   0    0    nan │
│                     coq-vst │ 1045.16  1052.58  -0.70 │  4767588637310   4800746683221  -0.69 │  7798562800789   7798195375245   0.00 │ 2069540  2071284  -0.08 │   0    0    nan │
│                coq-rewriter │  370.12   372.37  -0.60 │  1685283826663   1696113403196  -0.64 │  2775743153593   2775727277498   0.00 │ 1062792  1062448   0.03 │   0    0    nan │
│                    coq-hott │  160.44   161.25  -0.50 │   727467455005    732390970250  -0.67 │  1150006551134   1150054329181  -0.00 │  587120   586688   0.07 │   0    0    nan │
│  coq-performance-tests-lite │  772.02   775.76  -0.48 │  3505593301111   3522662809316  -0.48 │  5973419760850   5972963075592   0.01 │ 1824472  1824928  -0.02 │   0    0    nan │
│             coq-fiat-crypto │ 2790.99  2804.30  -0.47 │ 12695332756795  12760553924261  -0.51 │ 23306980759605  23303919866192   0.01 │ 3850456  3850132   0.01 │   0    0    nan │
│          coq-mathcomp-field │  111.93   112.45  -0.46 │   512103759428    514633177000  -0.49 │   837916804720    837848074849   0.01 │  653980   654072  -0.01 │   0    0    nan │
│            coq-math-classes │   90.16    90.54  -0.42 │   408409556784    410572915746  -0.53 │   572051401988    572077254131  -0.00 │  496596   495980   0.12 │   0    0    nan │
│              coq-verdi-raft │  559.41   561.43  -0.36 │  2553433210357   2564406286539  -0.43 │  3935326826387   3935086203398   0.01 │ 1228972  1229560  -0.05 │   0    0    nan │
│                coq-compcert │  279.52   280.29  -0.27 │  1267883725762   1272486685401  -0.36 │  1885420288801   1885381612349   0.00 │ 1099944  1098908   0.09 │   0    0    nan │
│                    coq-core │   97.74    98.00  -0.27 │   393812778549    392451076050   0.35 │   440908279247    440917363528  -0.00 │  274680   274608   0.03 │   1    0    nan │
│ coq-rewriter-perf-SuperFast │  920.08   922.45  -0.26 │  4196755238318   4210734330931  -0.33 │  7300708227506   7299458091612   0.02 │ 1145532  1145448   0.01 │   0    0    nan │
│                 coq-unimath │ 3927.08  3932.08  -0.13 │ 17964302256999  17986265097752  -0.12 │ 35844326403317  35844377935589  -0.00 │ 3323880  3322608   0.04 │   0    0    nan │
│                 coq-bignums │   27.85    27.87  -0.07 │   126923724856    127132951917  -0.16 │   173311844572    173075362307   0.14 │  466056   466392  -0.07 │   0    0    nan │
│                  coq-stdlib │  402.45   402.35   0.02 │  1657469830051   1646613301447   0.66 │  1413241027648   1413673251922  -0.03 │  609140   605200   0.65 │   0    0    nan │
│                  coq-geocoq │  691.09   688.90   0.32 │  3149712238355   3143038888993   0.21 │  4954663172576   4954506092350   0.00 │ 1075836  1075728   0.01 │   0    0    nan │
│               coq-fourcolor │ 1468.69  1462.92   0.39 │  6695821094069   6671239905114   0.37 │ 12086831016770  12086946703087  -0.00 │  706472   705160   0.19 │   0    0    nan │
│                  coq-flocq3 │   76.74    76.35   0.51 │   348403467492    348000605425   0.12 │   457436452501    456081136114   0.30 │  987768   988332  -0.06 │   0    0    nan │
│               coq-fiat-core │    0.65     0.63   3.17 │     2[1728](https://gitlab.com/coq/coq/-/jobs/2060215599#L1728)82477      2171983890   0.04 │     3708300202      3728236688  -0.53 │   77092    76792   0.39 │   2    0    nan │
│            coq-fiat-parsers │    0.63     0.61   3.28 │     2188385716      2166951966   0.99 │     3740434060      3727507302   0.35 │   77040    77048  -0.01 │   0    0    nan │
└─────────────────────────────┴─────────────────────────┴───────────────────────────────────────┴───────────────────────────────────────┴─────────────────────────┴─────────────────┘
```
For some reason it looks faster. Go figure.

**Comment by Alizter** [comment]
@herbelin ping, can you take a look at my comment after which I will merge.

File: doc/changelog/04-tactics/15634-master+fix-15606-clearbody-transitive-check.rst (hunk 1 above)
Comment (herbelin) [reply doc/changelog/04-tactics/15634-master+fix-15606-clearbody-transitive-check.rst]: Done, thanks.

**Comment by herbelin** [comment]
> For some reason it looks faster. Go figure.

It does a priori more checks because of possible dependencies but it is also more efficient in that it does not look in the prefix of the context before the oldest variable to clear, so, it saves a couple of `occur_var_in_decl`.

**Comment by Alizter** [comment]
merging soon

File: tactics/tactics.ml [hunk 2]
```diff
@@ -2096,60 +2100,62 @@ let check_decl env sigma decl =
...
-        let msg = match where with
-        | None -> str "Conclusion" ++ on_the_bodies ids
-        | Some id -> str "Hypothesis " ++ Id.print id ++ on_the_bodies ids
-        in
-        Tacticals.tclZEROMSG ~info msg
+          let env, sigma, ids = fold ids ctx in
+          if Id.Set.exists (fun id -> occur_var_in_decl env sigma id decl) ids then
+            let sigma = check_decl env sigma idl ids decl in (* can sigma really change? *)
+            let ids = Id.Set.add (get_id decl) ids in
+            push_named decl env, sigma, Id.Set.add (get_id decl) ids
```
Comment (JasonGross) [inline tactics/tactics.ml:2140]: Is there a reason that you add `get_id decl` to `ids` on both this line and the previous one?

Comment (herbelin) [reply tactics/tactics.ml:2140]: Some overlook. Shall we do a quick PR to remove the confusion?

**Pull request merged**